core/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java

```diff
@@ -438,18 +438,33 @@ public boolean hasAliases(final String[] aliases, String[] concreteIndices) {
 
     /**
      * Returns indexing routing for the given index.
+     * If routing is not null then the value is returned, parent routing and alias routing are ignored.
+     * If routing is null and there is an alias routing then the value is returned and parent is ignored.
+     * If routing is null and a parent is defined then the parent is returned (unless an alias routing is defined, see above).
      */
     // TODO: This can be moved to IndexNameExpressionResolver too, but this means that we will support wildcards and other expressions
     // in the index,bulk,update and delete apis.
-    public String resolveIndexRouting(@Nullable String routing, String aliasOrIndex) {
-        if (aliasOrIndex == null) {
+    public String resolveIndexRouting(@Nullable String parent, @Nullable String routing, String aliasOrIndex) {
+        if (routing != null) {
+            // The routing is explicit, ignore the alias and parent routing (if any).
```

Comment:
I agree that being lenient is bad but what about being too strict ;). IMO it's not a question of being lenient or not but the question is how the rules for routing are applied and in which order. If we want to keep the same behavior then there would be no way to override a parent routing or an alias routing explicitly. I am fine with that but what do you think of the  workaround for #3068 ? I could alos change this PR to handle the parent routing vs alias routing only. This would mean that if an alias routing is defined for the request then the parent routing is ignored. @rjernst @jpountz which one do you prefer ?